File: taesunpark/2021/1108~1114/입국 심사.py
def solution(n, times):
    answer = 0
    start = 1
    end = max(times) * n # 최악의 경우 모든 사람이 젤 느린 곳으로 가서 시간이 늦어지는 경우

    while start <= end:
        mid = (start + end) // 2
        people = 0
        for i in times:
            people += mid // i
            if people >= n:     # 최대 사람 수가 넘을 경우
                break

        if people >= n:
            answer = mid
            end = mid - 1
        else:
            start = mid + 1

    return answer

File: hyeon-uk/2021/1108~1114/입국심사.cpp
#include <string>
#include <vector>
#include<algorithm>
#include<iostream>
using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    vector<long long> temp;
    sort(times.begin(),times.end());
    long long left=1;
    long long right=n*(long long)times[times.size()-1];
    
    
    while(left<=right){
        long long sum=0;
        long long middle=(left+right)/2;
        for(int i=0;i<times.size();i++){
            sum+=middle/times[i];
        }
        
        if(sum<n){
            left=middle+1;
        }
        else{
            if(middle<=right) answer=middle;
            right=middle-1;
        }
    }
    return answer;
}
